R/list-transpose.R

```diff
@@ -0,0 +1,267 @@
+    x <- unclass(x)
+  }
+
+  # We disallow `NULL` elements. These would break `vec_size()` invariants of
+  # `list_transpose()` if we simply drop them via `list_interleave()`.
+  #
+  # Either `list_check_all_vectors()` errors, or the user supplied `null` which
+  # will replace `NULL`s with size 1 vectors before we `list_interleave()`.
+  #
+  # For example:
+  #
+  # ```
+  # list_transpose(list(1:4, NULL, 5:8))
+  # ```
+  #
+  # Input:
+  # - List size 3
+  # - Element size 4
+  # Output:
+  # - List size 4
+  # - Element size 3
+  #
+  # But if we drop `NULL` you'd get:
+  # - List size 4
+  # - Element size 2
+  #
+  # Users should instead use `null` to replace `NULL` elements with something
+  # else, like `NA`. This is similar to `purrr::list_transpose(default =)` and
+  # `keep_empty` in some tidyr functions.
+  #
+  # ```
+  # list_transpose(list(1:4, NULL, 5:8), null = NA)
+  # ```
+  allow_null <- !is_null(null)
+
+  list_check_all_vectors(
+    x,
+    allow_null = allow_null,
+    arg = x_arg,
+    call = error_call
+  )
+
+  ptype <- list_transpose_ptype_common(
+    x,
+    null,
+    ptype,
+    x_arg,
+    error_call
+  )
+
+  if (!is_null(null)) {
+    # Do `null` checks regardless of usage
+    null <- vec_cast(
+      x = null,
+      to = ptype,
+      x_arg = "null",
+      to_arg = "",
+      call = error_call
+    )
+    vec_check_size(
+      null,
+      size = 1L,
+      arg = "null",
+      call = error_call
+    )
+
+    if (vec_any_missing(x)) {
+      null <- list(null)
+      x <- vec_assign(x, vec_detect_missing(x), null)
+    }
+  }
+
+  flat <- list_interleave(
+    x,
+    size = size,
+    ptype = ptype,
+    name_spec = "inner",
+    x_arg = x_arg,
+    error_call = error_call
+  )
+
+  x_size <- vec_size(x)
+
+  if (is_null(size)) {
+    # Identical to `elt_size <- vec_size_common(!!!x)`, but faster.
+    # Utilizes known info about the `list_interleave()` return value.
+    if (x_size == 0L) {
+      elt_size <- 0L
+    } else {
+      elt_size <- vec_size(flat) / x_size
+    }
+  } else {
+    elt_size <- size
+  }
+
+  sizes <- vec_rep(x_size, times = elt_size)
+
+  # Chop the one big vector into transposed pieces of size `x_size`
+  out <- vec_chop(flat, sizes = sizes)
+
+  out
+}
+
+# Computes the `ptype` incorporating both `x` and `null`
+#
+# Like `ptype_finalize()` in `vec_recode_values()` and `vec_if_else()`
+list_transpose_ptype_common <- function(
+  x,
+  null,
+  ptype,
+  x_arg,
+  error_call
+) {
+  if (!is_null(ptype)) {
+    # Validate and return user specified `ptype`
+    ptype <- vec_ptype(ptype, x_arg = "ptype", call = error_call)
+    return(vec_ptype_finalise(ptype))
+  }
+
+  # Compute from `x`
+  ptype <- vec_ptype_common(!!!x, .arg = x_arg, .call = error_call)
+
+  if (!is_null(null)) {
+    # Layer in `null`
+    ptype <- vec_ptype2(
+      x = null,
+      y = ptype,
+      x_arg = "null",
+      y_arg = "",
+      call = error_call
+    )
+  }
+
+  ptype
+}
```

Comment:
Same trick used in `vec_recode_values()` and `vec_if_else()`

Basically I really want to avoid `vec_ptype_common(!!!x, null = null)` because that materializes a new list, and if `x` is really long then that is needlessly expensive.